### PR TITLE
Pandas->DuckDB Series of dtype='O' conversion

### PR DESCRIPTION
**Pull request by Tishj**
This PR changes conversion of Pandas Series objects of dtype 'object' to the right DuckDB Logical+Physical Type
Previously these would be converted to string, losing all type specific properties in the process.

This now lets us preserve types like datetime, lists and dictionaries.

NaN ('float') is converted to SQLNULL if encountered in an object column
Cases like [True, False, NaN] would turn into [1.0, 0.0, NaN] without this

Dictionary is a bit of a special case, as it can turn into both STRUCT and MAP.
We try to preserve the intentional format as much as possible, meaning that a dictionary will result in a STRUCT if:
- All keys are strings or convertible to a string (have `__str__` defined)
- Across all rows, each struct has the same amount of fields, and the keys are case-insensitively identical, as well as the types.  

And it will result in a MAP if:
- There are only 2 key-value pairs present (keys are ignored in this case)
- There are 0 key-value pairs (the dictionary is empty)
- Both values are lists, and those lists are of the same size  

Since STRUCT is more restrictive than MAP, we will attempt to fall back to MAP when STRUCT fails before failing entirely if all fields of the struct are of the same type, or the types are compatible.
In this case the type of the keys of the resulting MAP is set to VARCHAR. (I could probably preserve the actual type of the keys if necessary but for now I haven't because it would require a bit of refactoring)

Decimal values in python (`decimal.Decimal`) support `inf` and `nan` which DECIMAL does not, so in those instances we will result to DOUBLE

### CODE REVIEW

File: tools/pythonpkg/src/python_conversion.cpp
Comment (Mytherin) [inline tools/pythonpkg/src/python_conversion.cpp:76]: I wonder about the performance cost of doing all of these imports once per value - can we cache these somewhere (e.g. in the connection)? Python caches the actual modules but this will still likely be 10+ dictionary lookups per value which could be very significant.

Comment (Tishj) [reply tools/pythonpkg/src/python_conversion.cpp:76]: But I guess since we know which modules+types we're looking for, we can probably do something smarter that doesn't require so many lookups

**Comment by Tishj** [comment]
> You still need the analyze_df stuff? I thought this was dropped and the detection cold would always run on the binding phase, no?

Yes that's the plan, I marked this PR as a draft for now, it's passing tests but it's far from clean enough to get reviewed :)

File: tools/pythonpkg/src/python_conversion.cpp
Comment (Mytherin) [reply tools/pythonpkg/src/python_conversion.cpp:76]: I would go with a class that caches a fixed set of imports, e.g. something like this:

```cpp
struct PyImportCache {
  py::obj datetime_mod;
  py::obj datetime_date;
  py::obj datetime_datetime;
  py::obj datetime_time;
  py::obj decimal_mod;
  py::obj decimal_decimal;
  py::obj numpy_mod;
  py::obj numpy_ndarray;
  py::obj uuid_mod;
  py::obj uuid_uuid;
};
```

I think this is useful elsewhere in the Python API too. Then we can extend it as required with new cached imports, and no string lookups are required at all.

**Comment by Tishj** [comment]
Hmm I just found a bit of a discrepancy between Python's `datetime.time` and DuckDB's `Value::TIME` 

`datetime.time` records up to nanoseconds, but our TIME value only goes up to microseconds so we can't record this type losslessly

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Mause) [inline tools/pythonpkg/src/pandas_analyzer.cpp:115]: Seems strange to directly call the dunder method explicitly here, does pybind not have something higher level that uses the index protocol under the hood?

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Tishj) [reply tools/pythonpkg/src/pandas_analyzer.cpp:200]: I've explained this in the description of the PR, but in short: to explicitly convert dict->map, both values of the dict need to be lists; one for the maps keys, and one for the maps values

Comment (Tishj) [reply tools/pythonpkg/src/pandas_analyzer.cpp:200]: If you supply key-value pairs, struct will be considered first, and only after the constraints of struct aren't respected do we convert it to a map

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Tishj) [reply tools/pythonpkg/src/pandas_analyzer.cpp:115]: Hmm maybe, I hadn't considered this. But looking just now, it doesnt look like pybind11 has such a thing.

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Mause) [reply tools/pythonpkg/src/pandas_analyzer.cpp:200]: Wouldn't that be dependent on the order of the values in the dictionary then? Which isn't guaranteed on older versions of python? Unless of course you're fixing that elsewhere, this is a large pr 😅

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Mause) [reply tools/pythonpkg/src/pandas_analyzer.cpp:115]: I imagine the lookup cost for each call wouldn't be great either :/

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Tishj) [reply tools/pythonpkg/src/pandas_analyzer.cpp:200]: Hmm I hadn't considered that, you mean that the order of the key-value pairs is undefined right? Yea that's an oversight and definitely something I have to look into

File: tools/pythonpkg/src/vector_conversion.cpp
Comment (Mause) [inline tools/pythonpkg/src/vector_conversion.cpp:157]: One of the biggest issues with the python client at present is the widespread use of `runtime_error` within the codebase (which is mapped straight to the python  `RuntimeError`), as it means all exception handling requires the use of string matching on the exception message in duckdb-engine and I imagine other consumers of the api. Would it be possible to fix this before the pr is merged? Ideally there would be a root "DuckDBException" that could be cleanly exposed via pybind11

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/src/vector_conversion.cpp
Comment (Mytherin) [reply tools/pythonpkg/src/vector_conversion.cpp:157]: I would say this is out of scope for this PR and should be fixed in a separate one

**Comment by Tishj** [comment]
I've looked into the failing regression test, which is `pandas_tpch` that normally runs in 4.9 seconds, and now runs in approximately 49 seconds.

As I suspected, this is caused by tpch containing a LOT of object columns, which (as far as I can tell) only contain strings.
I've set the default minimum to 100 and the default percentage to 10. Which causes a lot of objects to be analyzed in the case of tpch.

**Comment by Tishj** [comment]
Currently this test in `tools/pythonpkg/tests/fast/test_multithread.py` hangs infinitely (probably deadlocked) on Windows
```py
    def test_df_append(self, duckdb_cursor):
        duck_threads = DuckDBThreaded(10,df_append)
        duck_threads.multithread_test()
```
I first thought it might be due to my dodgy code in the construction of the Python Import Cache, but after fixing that the issue still persists

**Comment by Mytherin** [comment]
Thanks for all the debugging work!

So usually a dead-lock happens because of lock inversion. That is to say, if we have two threads `T1` and `T2`, and two locks `L1` and `L2`, the following locking order leads to a dead-lock:

```
T1: LOCK(L1) -> LOCK(L2)
T2: LOCK(L2) -> LOCK(L1)
```

Because `T1` will grab `L1` and wait for `L2`, and `T2` will grab `L2` and wait for `L1`. This causes a deadlock, as both threads are waiting for the other thread to release the lock, which will never happen. 

In this case, there are three locks at play:

* The `py_connection_lock` in the connection
* The client context lock
* The GIL

### Analyzing the function

Looking at the functions that are called in the `DuckDBPyConnection::Append`, we have `RegisterPythonObject` and `Execute`.

When we enter this function we hold the GIL.

In `RegisterPythonObject` we release the GIL, take the client context lock, and then run the Pandas Bind which takes the GIL again.

* Lock order: `UNLOCK(GIL) -> LOCK(CLIENT_CONTEXT) -> LOCK(GIL)`

In `Execute` we unlock the GIL, lock the `py_connection_lock`, lock the GIL again, lock the client context, and then run the Pandas Bind which takes the GIL again.

* Lock order: `UNLOCK(GIL) -> LOCK(py_connection_lock) -> LOCK(GIL) -> LOCK(CLIENT_CONTEXT) -> LOCK(GIL)`

We can see the lock order inversion here:

* RegisterPythonObject: `LOCK(CLIENT_CONTEXT) -> LOCK(GIL)`
* Execute: `LOCK(GIL) -> LOCK(CLIENT_CONTEXT)`

The mistake here seems to be that `AcquireConnectionLock` re-acquires the GIL, which it should not. Instead, the GIL should stay released. Could you perhaps try to replace this line:

```cpp
auto cur_py_connection_lock = AcquireConnectionLock();
```

With the following code:

```cpp
py::gil_scoped_release release;
unique_lock<std::mutex> lock(py_connection_lock);
```

And check if that resolves the problem?

**Comment by Tishj** [comment]
Applied the suggested fix, but on a Windows instance this still fails (test 12 is the `test_df_append`)
image

**Comment by Tishj** [comment]
I've narrowed down the issue to one file, maybe even just one function
It's BindPandas, looks like it's caused by either
1. The optimization I put in to use `num` from dtype instead of converting to string and doing string comparisons
2. The reordering of operations performed to take the logical type, numpy_col and mask of a dataframe series

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Mytherin) [inline tools/pythonpkg/src/pandas_analyzer.cpp:25]: `ENUM` is actually not nested as well, and is after `LIST` in the logical type enum. Perhaps we could do a comparison to the various nested types in a separate function (e.g. `TypeIsNested`)?

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Mytherin) [reply tools/pythonpkg/src/pandas_analyzer.cpp:122]: The `py::handle` seems to have the `[]` operator overloaded to call [`__getitem__`](https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L89), perhaps that can be used?

The lookup cost indeed will be slow but that is not a big problem as this only happens during the bind.

File: tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:370]: Could we add tests with:

*A large data frame where the sampling picks up the incorrect type (e.g. we have 10000 times the element `1` and 1 time the element `[1, 2, 3]`)
* Many mixed objects (e.g. `1, [1], {'a': 42}, 1.0`)
* Mixed numeric objects (e.g. `[1, None, 1.0, decimal.Decimal(1.0)]`)
* More recursive nested types (e.g. lists of structs, structs containing lists, including nulls at various levels), e.g.:


```py
# lists of lists
numpy.array([
	[[1, 2, 3], None, [None, 4, 5]],
	None,
	[[6, None, 8], [9], [], None]
], dtype=object)

# structs with lists
numpy.array([
	{'a': [1, None, 3], 'b': 42},
	None,
	{'a': None, 'b': None},
	{'a': [4, None, 6], 'b': None}
], dtype=object)
```

**Review by Mytherin — COMMENTED** [review]
Thanks for all the updates, and well done on fixing the deadlock! Looks great. Some more comments from my side below, perhaps @pdet can have another look at the PR as well.

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Tishj) [reply tools/pythonpkg/src/pandas_analyzer.cpp:25]: Yea good idea, JSON is also a nested type, no?

So the list of nested types are:
- STRUCT
- LIST
- MAP
- JSON
- AGGREGATE_STATE (?)
- TABLE

Is that list correct, am I missing any, or are some actually not nested? (I'm not sure about AGGREGATE_STATE)

File: tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py
Comment (Tishj) [reply tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:370]: py::handle only overloads [] for `const char*` and `py::handle` so the closest I can get it is this:
`auto next_item_type = GetItemType(column[py::int_(i)], can_convert);`

Which I'm not super excited about

File: tools/pythonpkg/src/pandas_analyzer.cpp
Comment (Mytherin) [reply tools/pythonpkg/src/pandas_analyzer.cpp]: It is only STRUCT, LIST and MAP. TABLE is unused, and aggregate_state is a blob.

File: tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py
Comment (Mytherin) [reply tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:370]: That solution looks good to me.

Comment (Tishj) [reply tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:370]: I will add more tests with recursive nested types, the only problem I have there is creating the reference tables. I wrote something (pretty horrible) for this when I was writing tests for testing arrow list -> duckdb, only problem is it uses TransformPythonValue (for the (?) resolution) and that's also what is used to transform the values of the panda dataframe columns.

```py
def create_and_register_comparison_result(column_list, duckdb_conn):
    columns = ",".join([f'{name} {dtype}' for (name, dtype, _) in column_list])
    column_amount = len(column_list)
    assert(column_amount)
    row_amount = len(column_list[0][2])
    inserted_values = []
    for row in range(row_amount):
        for col in range(column_amount):
            inserted_values.append(transform(column_list[col][2][row]))
    inserted_values = tuple(inserted_values)

    column_format = ",".join(['?' for _ in range(column_amount)])
    row_format = ",".join([f"({column_format})" for _ in range(row_amount)])
    query = f"""CREATE TABLE test ({columns});
        INSERT INTO test VALUES {row_format};
    """
    duckdb_conn.execute(query, inserted_values)
  ```

Comment (Mytherin) [reply tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py:370]: How about just using SQL:

create table tbl as select [[1,2,3],[4,5]] col union all select null;

**Comment by Tishj** [comment]
To properly test conversion from decimals, I feel like issue #4106 needs to be fixed first, so I've added a test that passes as long as the bug is still there. In the hopes that it will bring attention to this lack of tests for this conversion when it's fixed
(And this comment references it, which hopefully also helps)

**Review by Mytherin — COMMENTED** [review]
Thanks for the fixes! Looks excellent. Some more minor comments, otherwise I think this is good to go:

File: tools/pythonpkg/setup.py
Comment (Tishj) [reply tools/pythonpkg/setup.py:161]: Ah good idea, I had found this list in there, but wasn't sure of how to get it cleanly into both

File: src/common/types/interval.cpp
Comment (Tishj) [reply src/common/types/interval.cpp]:
```c++
interval_t Interval::Invert(interval_t interval) {
	interval.days = -interval.days;
	interval.micros = -interval.micros;
	interval.months = -interval.months;
	return interval;
}
```
would this suffice, or do you also want me to unify this with subtract.cpp?

File: src/common/types/vector.cpp
Comment (Tishj) [reply src/common/types/vector.cpp:308]: I figured that would be more expensive to do for every SetValue call, so I only added it for the case I knew was problematic, but I'll change it

File: src/common/types/interval.cpp
Comment (Mytherin) [reply src/common/types/interval.cpp]: Perhaps subtract.cpp can call this method as well?

File: src/common/types/vector.cpp
Comment (Tishj) [reply src/common/types/vector.cpp:308]: Hmm because of this change a test is failing
```c++
	result.emplace_back(LogicalType::BLOB, "blob", Value::BLOB("thisisalongblob\\x00withnullbytes"),
	                    Value("\\x00\\x00\\x00a"));
```
The last value was previously not being cast, because VARCHAR and BLOB share the same internal type
But now that we are casting them, this no longer remains escaped
Previous result:
```py
[(b'thisisalongblob\x00withnullbytes',), (b'\\x00\\x00\\x00a',), (None,)]
```
New result
```py
[(b'thisisalongblob\x00withnullbytes',), (b'\x00\x00\x00a',), (None,)]
```

Comment (Mytherin) [reply src/common/types/vector.cpp:308]: Looks to me like an oversight actually, that should be `Value::BLOB` there as well I believe

Comment (Tishj) [reply src/common/types/vector.cpp:308]: I'm pretty sure this is a bug
`\\` should get escaped down to `\`, resulting in the contents of the string being:
`['\', '\0', '\', '\0', '\', '\0', 'a']`
If we convert that back to BLOB every character should be interpreted separately, resulting in the same contents (so essentially no cast)

The issue probably stems from using the same cast for string values from the queries as the VARCHAR values inside the database
```sql
duckdb> select '\x00\x00\x00a'::BLOB;
┌───────────────────────┐
│ '\x00\x00\x00a'::BLOB │
╞═══════════════════════╡
│ 00000061              │
└───────────────────────┘
```

Comment (Mytherin) [reply src/common/types/vector.cpp:308]: This is not a bug but intended behavior, for blobs `\x00` is an escape code that means `the byte 0`. The resulting blob in the example (`\x00\x00\x00a`) should be 4 bytes, `[0, 0, 0, 97]`. Previously the string `\x00\x00\x00a` was stored as a string, but silently converted to a blob because of the missing type cast. This caused the contents of the blob to be the ascii values of that string instead of handling the escape code, i.e. the blob used to have 13 characters with as content `[92, 120, 48, 48, 92, 120, 48, 48, 92, 120, 48, 48, 97]`. The new result seems correct to me as what was intended was likely to use the hex escape codes to create a blob with null bytes.

Comment (Mytherin) [reply src/common/types/vector.cpp:308]: Ah, I see the confusion now, the escaping double back-slash is to escape the C++ string. The C++ string `"\\x00" gets converted into the actual string "\x00". In order to escape the blob a double escaped backslash would be necessary (i.e. "\\\\x00").

Comment (Tishj) [reply src/common/types/vector.cpp]: I did find another issue possibly caused by this
Observing a crash in `test_alias_table`
Appears to be an infinite recursion, looks like it's caused by this change
```
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010c10633c libduckdb.dylib`duckdb::Vector::SetValue(this=0x0000600003700960, index=0, val=0x000000016fdf7df0) at vector.cpp:302:12
   299                  return child.SetValue(sel_vector.get_index(index), val);
   300          }
   301          if (val.type() != GetType()) {
-> 302                  SetValue(index, val.CastAs(GetType()));
   303                  return;
   304          }
   305 
(lldb) p val.type()
(const duckdb::LogicalType) $4 = {
  id_ = VARCHAR
  physical_type_ = VARCHAR
  type_info_ = nullptr {
    __ptr_ = nullptr
  }
}
(lldb) p GetType()
(const duckdb::LogicalType) $5 = {
  id_ = VARCHAR
  physical_type_ = VARCHAR
  type_info_ = std::__1::shared_ptr<duckdb::ExtraTypeInfo>::element_type @ 0x000060000211c798 strong=19 weak=1 {
    __ptr_ = 0x000060000211c798
  }
}
(lldb) n
Process 84474 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010c10633c libduckdb.dylib`duckdb::Vector::SetValue(this=0x0000600003700960, index=0, val=0x000000016fdf7b50) at vector.cpp:302:12
   299                  return child.SetValue(sel_vector.get_index(index), val);
   300          }
   301          if (val.type() != GetType()) {
-> 302                  SetValue(index, val.CastAs(GetType()));
   303                  return;
   304          }
   305 
(lldb) p GetType()
(const duckdb::LogicalType) $6 = {
  id_ = VARCHAR
  physical_type_ = VARCHAR
  type_info_ = std::__1::shared_ptr<duckdb::ExtraTypeInfo>::element_type @ 0x000060000211c798 strong=19 weak=1 {
    __ptr_ = 0x000060000211c798
  }
}
(lldb) p val.type()
(const duckdb::LogicalType) $7 = {
  id_ = VARCHAR
  physical_type_ = VARCHAR
  type_info_ = nullptr {
    __ptr_ = nullptr
  }
}
```

Comment (Mytherin) [reply src/common/types/vector.cpp]: Ah, this is caused by a recent PR. The problem seems to be that the varchar type has an alias, but `Vector::GetValue` does not correctly set the alias of the value. Try rewriting `Vector::GetValue` to the following:

```cpp
Value Vector::GetValue(idx_t index) const {
    auto result = GetValue(*this, index);
    result.type().SetAlias(GetType().GetAlias());
    return result;
}
```

File: tools/pythonpkg/setup.py
Comment (Tishj) [reply tools/pythonpkg/setup.py]: Update: setup.py is not a fan (as in: it won't let you) of bundling files outside of the root directory, even through MANIFEST.in

So I would have to move package_build.py to tools/pythonpkg and also python_helpers.py because package_build depends on that.
Which would result in having to rewrite a bunch of code in amalgamation.py, setup.py and package_build + possibly others, because they rely on these files being in certain locations for relative paths. Because of this I think this should be its own PR.

Comment (Mytherin) [reply tools/pythonpkg/setup.py]: Agreed, let's revert to the original

Comment (Tishj) [reply tools/pythonpkg/setup.py]: Hmm after looking at it more closely, it's only needed in a less broad scope as where I was including it, it seems to work now ?

**Comment by Mytherin** [comment]
Thanks for all the work! Looks great.

**Pull request merged**
